lib/Server.js

```diff
@@ -335,11 +343,13 @@ function Server(compiler, options) {
     },
 
     setup: () => {
+      log('Using "setup" is deprecated and will be removed in the next major version. Please use the "before" and "after" hooks instead.');
+      log('If "setup" was working fine for you until now, simply replace it with "before"');
       if (typeof options.setup === 'function') { options.setup(app, this); }
     }
   };
 
-  const defaultFeatures = ['setup', 'headers', 'middleware'];
+  const defaultFeatures = ['before', 'setup', 'headers', 'middleware'];
```

Comment:
@shellscape @mrtnbroder doesn't setup being in this array mean that everyone will see the deprecation warning no matter what? Shouldn't it not be in there but one of the things below this say `if(options.setup) defaultFeatures.push('setup')` or something? (or maybe keep it in but if options.setup is absent then remove it).

It seems that the loop on line 367 will always run the deprecation warning. 

I'm asking because when i upgraded to 2.9 i noticed the deprecation warning even though I don't use this hook at all (though it may just be that a dependency of mine needs upgrading. I'm not totally sure)

Reply:
either that or putting the logs on lines 346-347 inside the if on line 348

Reply:
I made #1120 for this